runtime/bindings/python/loop.cc

```diff
@@ -0,0 +1,254 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "./loop.h"
+
+#include <cstdio>
+#include <vector>
+
+#include "./hal.h"
+#include "iree/base/internal/synchronization.h"
+
+namespace iree::python {
+
+namespace {
+
+static const char kHalDeviceLoopDocstring[] =
+    R"(Loop for handling events/timepoints associated with a single device.
+
+In the single device case, this is the best option for running a host event
+loop for dispatching callbacks upon satisfaction of semaphores and fences as
+it just calls down into the driver and does not involve complicated
+choreography to join across devices, etc.
+
+This is not a general purpose asyncio.loop but it can interop with asyncio
+by satisfying futures derived from a single device. It should generally be
+launched on a background thread in such cases.
+)";
+
+class HalDeviceLoop {
+ public:
+  HalDeviceLoop(HalDevice device, size_t initial_capacity)
+      : device_(std::move(device)) {
+    iree_slim_mutex_initialize(&mu_);
+    CheckApiStatus(
+        iree_hal_semaphore_create(device_.raw_ptr(), 0, &control_sem_),
+        "create semaphore");
+
+    next_pending_futures_.reserve(initial_capacity);
+  }
+  ~HalDeviceLoop() {
+    // Cancel all futures.
+    iree_slim_mutex_lock(&mu_);
+    for (auto &entry : next_pending_futures_) {
+      py::handle future = std::get<2>(entry);
+      CancelFuture(future);
+    }
+    next_pending_futures_.clear();
+    iree_slim_mutex_unlock(&mu_);
+
+    iree_slim_mutex_deinitialize(&mu_);
+    iree_hal_semaphore_release(control_sem_);
+  }
+
+  void Run() {
+    py::gil_scoped_release gil_release;
+    // Wait list.
+    size_t wait_capacity = 0;
+    iree_hal_semaphore_t **wait_semaphores = nullptr;
```

Comment:
since you use vector in here you can use it for these two as well - .data() will give you back the semaphores/payloads that you can pass to the C functions - will let you use resize when growing and ignore the free and such